src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/CompiledPageRouteModelProvider.cs

```diff
@@ -58,26 +69,63 @@ private void EnsureCache()
+
                 var cachedApplicationModels = new List<PageRouteModel>();
                 foreach (var viewDescriptor in GetViewDescriptors(_applicationManager))
                 {
-                    if (!viewDescriptor.RelativePath.StartsWith(rootDirectory, StringComparison.OrdinalIgnoreCase))
+                    PageRouteModel model = null;
+                    if (viewDescriptor.RelativePath.StartsWith(rootDirectory, StringComparison.OrdinalIgnoreCase))
                     {
-                        continue;
+                        model = GetPageRouteModel(rootDirectory, viewDescriptor);
+                    }
+                    else if (_pagesOptions.EnableAreas && viewDescriptor.RelativePath.StartsWith(areaRootDirectory, StringComparison.OrdinalIgnoreCase))
+                    {
+                        model = GetAreaPageRouteModel(areaRootDirectory, viewDescriptor);
                     }
 
-                    var viewEnginePath = GetViewEnginePath(rootDirectory, viewDescriptor.RelativePath);
-                    var model = new PageRouteModel(viewDescriptor.RelativePath, viewEnginePath);
-                    var pageAttribute = (RazorPageAttribute)viewDescriptor.ViewAttribute;
-                    PageSelectorModel.PopulateDefaults(model, pageAttribute.RouteTemplate);
-
-                    cachedApplicationModels.Add(model);
+                    if (model != null)
+                    {
+                        cachedApplicationModels.Add(model);
+                    }
                 }
 
                 _cachedModels = cachedApplicationModels;
             }
         }
 
+        private PageRouteModel GetPageRouteModel(string rootDirectory, CompiledViewDescriptor viewDescriptor)
+        {
+            var viewEnginePath = GetRootTrimmedPath(rootDirectory, viewDescriptor.RelativePath);
+            var model = new PageRouteModel(viewDescriptor.RelativePath, viewEnginePath);
+            var pageAttribute = (RazorPageAttribute)viewDescriptor.ViewAttribute;
+            PageSelectorModel.PopulateDefaults(model, viewEnginePath, pageAttribute.RouteTemplate);
+            return model;
+        }
+
+        private PageRouteModel GetAreaPageRouteModel(string areaRootDirectory, CompiledViewDescriptor viewDescriptor)
+        {
+            var rootTrimmedPath = GetRootTrimmedPath(areaRootDirectory, viewDescriptor.RelativePath);
+
+            if (PageSelectorModel.TryParseAreaPath(_pagesOptions, rootTrimmedPath, _logger, out var result))
+            {
+                var model = new PageRouteModel(viewDescriptor.RelativePath, result.viewEnginePath)
+                {
+                    RouteValues = { ["area"] = result.areaName },
+                };
+                var pageAttribute = (RazorPageAttribute)viewDescriptor.ViewAttribute;
+                PageSelectorModel.PopulateDefaults(model, result.pageRoute, pageAttribute.RouteTemplate);
+
+                return model;
+            }
+
+            return null;
```

Comment:
What are some examples of this? I think it at least deserves a comment... Does it make sense for this to be an error? (I'm not sure what I think)

Reply:
Sure. Most of it is stray Razor Page cshtml sitting in the Areas root e.g. `/Areas/Page.cshtml`  or `/Areas/MyArea/Page.cshtml`. We log warning for these as part of `TryParseAreaPath` but I'm not sure if that's enough.